test/unit/maybe_callback.test.js

```diff
@@ -0,0 +1,196 @@
+const client = new mongodbLegacy.MongoClient(iLoveJs);
+const db = new mongodbLegacy.Db(client, 'animals');
+const collection = new mongodbLegacy.Collection(db, 'pets');
+const namespace = MongoDBNamespace.fromString('animals.pets');
+
+const OVERRIDDEN_CLASSES_GETTER = new Map([
+  ['Admin', () => new mongodbLegacy.Admin(db)],
+  ['FindCursor', () => new mongodbLegacy.FindCursor(client, namespace)],
+  ['ListCollectionsCursor', () => new mongodbLegacy.ListCollectionsCursor(db, {})],
+  ['ListIndexesCursor', () => new mongodbLegacy.ListIndexesCursor(collection)],
+  ['AggregationCursor', () => new mongodbLegacy.AggregationCursor(client, namespace)],
+  ['ChangeStream', () => new mongodbLegacy.ChangeStream(client)],
+  ['GridFSBucket', () => new mongodbLegacy.GridFSBucket(db)],
+  ['Collection', () => new mongodbLegacy.Collection(db, 'pets')],
+  ['Db', () => new mongodbLegacy.Db(client, 'animals')],
+  ['MongoClient', () => new mongodbLegacy.MongoClient(iLoveJs)]
+]);
+
+function* generateTests() {
+  for (const [className, getInstance] of OVERRIDDEN_CLASSES_GETTER) {
+    /** @type {{ className: string; method: string; returnType: string; special?: boolean; }[]} */
+    let methods = classesToMethods.get(className);
+    if (methods == null) methods = [];
+    for (const { method, special, possibleCallbackPositions } of methods) {
+      if (special === true) continue;
+      const apiName = `${className}.${method}`;
+      const instance = getInstance();
+      yield {
+        className,
+        method,
+        instance,
+        apiName,
+        possibleCallbackPositions,
+        makeStub: superPromise => {
+          return sinon.stub(mongodbDriver[className].prototype, method).returns(superPromise);
+        }
+      };
+    }
+  }
+}
+
+describe('Maybe Callback', () => {
+  afterEach(() => {
+    sinon.restore();
+  });
+
+  for (const {
+    apiName,
+    instance,
+    method,
+    possibleCallbackPositions,
+    makeStub
+  } of generateTests()) {
+    const functionLength = instance[method].length;
+
+    context(`${apiName}()`, () => {
+      for (const callbackPosition of possibleCallbackPositions != null
+        ? possibleCallbackPositions
+        : [1, 2]) {
+        if (callbackPosition === 2) {
+          if (functionLength < 2) continue;
+          if (
+            apiName.includes('forEach') ||
+            ['GridFSBucket.delete', 'GridFSBucket.rename', 'Db.dropCollection'].includes(apiName)
+          ) {
+            continue;
+          }
+        }
+        it(`returns void and uses callback(_, result) in position ${
+          functionLength - callbackPosition
+        }`, async () => {
+          const superPromise = Promise.resolve({ message: 'success!' });
+          makeStub(superPromise);
+
+          expect(instance).to.have.property(method).that.is.a('function');
+
+          const callback = sinon.spy();
+
+          const args = Array.from({ length: functionLength }, (_, i) => i);
```

Comment:
Can you explain what these lines are doing?  I can't wrap my mind around why or how this works.  Here's an example that I'm looking at for a contrived function with the signature `foo(options?, cb?)`, calling the cb position at the end

```
foo({}, cb)

functionLength - 2
cbPosition - 2
args at line 98 - [0, 1]
args at line 99 - [cb, 1]
args at line 100 - 2 - (2 - 1) = 1, so [cb]
```

so we're testing the overload `foo(cb)`

---

now calling the cb position at the start

```
foo(cb)

functionLength - 2
cbPosition - 1
args at line 98 - [0, 1]
args at line 99 - [0, cb]
args at line 100 - 2 - (1 - 1) = 2, so [0, cb]
```

A callback position of `n` corresponds to the test that calls the callback at `functionLength - n`, which is not particularly intuitive.  Can we not adjust the logic so that a callback position of `n` corresponds to the test that calls a callback at position `n`?